src/Storages/Statistics/Statistics.cpp

```diff
@@ -56,58 +78,77 @@ Float64 ColumnStatistics::estimateGreater(Float64 val) const
 
 Float64 ColumnStatistics::estimateEqual(Float64 val) const
 {
-    if (stats.contains(StatisticsType::Uniq) && stats.contains(StatisticsType::TDigest))
+    /// Special case for the unlikely case that the user created 'uniq' and 'tdigest' statistics.
+    for (const auto & stat : statistics)
     {
-        auto uniq_static = std::static_pointer_cast<UniqStatistics>(stats.at(StatisticsType::Uniq));
-        /// 2048 is the default number of buckets in TDigest. In this case, TDigest stores exactly one value (with many rows)
-        /// for every bucket.
-        if (uniq_static->getCardinality() < 2048)
+        auto statistics_uniq = std::dynamic_pointer_cast<StatisticsUniq>(stat);
+        if (statistics_uniq && statistics_uniq->estimateCardinality() < 2024)
         {
-            auto tdigest_static = std::static_pointer_cast<TDigestStatistics>(stats.at(StatisticsType::TDigest));
-            return tdigest_static->estimateEqual(val);
+            /// 2048 is the default number of buckets in TDigest. In this case, TDigest stores exactly one value (with many rows)
+            /// for every bucket.
+            for (const auto & stat2 : statistics)
+            {
+                auto statistics_tdigest = std::dynamic_pointer_cast<StatisticsTDigest>(stat2);
+                if (statistics_tdigest)
+                    return *(statistics_tdigest->estimateEqual(val));
+            }
         }
     }
+
+    for (const auto & stat : statistics)
+    {
+        /// Return the estimation of the first statistics object which provides an estimation.
+        /// TODO: In the (unlikely) case multiple statistics objects provide an estimation, we need some heuristics which
+        ///       returns the most accurate estimation (this necessitates some notion of "estimation quality").
+        auto estimation = stat->estimateEqual(val);
```

Comment:
My idea is not estimating equal from tdigest when bucket is greater than 2048, it will cause great bias.

Reply:
In order to support l. 86-96, `StatisticsTDigest` now implements `estimateEqual` which indeed means that TDigest potentially estimates equal in l. 97-103 which (as you said) can be biased when there are lots of buckets. As a remedy, we could change l. 104 to something like `if (estimation && statistics != TDigest)` but I generally think that special casing like this messes with the generality of the approach. In fact, in a good design, the base class should not enumerate or deal with concrete subclasses at all.

I would still like to advocate my original idea that each statistics type should be used for estimations it can robustly (= in all situations) support. That is:
- for TDigest: range predicates
- for Uniq: - currently not clear what it can be used for -
- for count-min sketch: equality predicates

If we document in which situations / for which predicates a certain statistics type can be used, users will have an easy time understanding if certain statistics will help with their queries. (By the way, we should also document the storage overhead for each statistics object, at least as a rule of thumb.)

Let me know your preference.

Reply:
I left my comment here explaining why statistics should cooperate with each other https://github.com/ClickHouse/ClickHouse/pull/65792#discussion_r1663177574

For sure the users need to which statistics is responsible to estimate which kind of estimation, and we could write `*** statistics type should be used for *** estimations`, but it also benificial that **when we don't have our favourite statistics** we can use some other statistics to estimate.

And idealy, users don't have to create statistics by themselves (Some users create histograms for `String` type), ClickHouse could know what kinds of statistics to build.